File: Android/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/loginfo/util/SearchCriteria.java
package com.didichuxing.doraemonkit.kit.loginfo.util;

import android.text.TextUtils;

import com.didichuxing.doraemonkit.kit.loginfo.LogLine;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class SearchCriteria {

    public static final String PID_KEYWORD = "pid:";
    public static final String TAG_KEYWORD = "tag:";

    private static final Pattern PID_PATTERN = Pattern.compile("pid:(\\d+)", Pattern.CASE_INSENSITIVE);
    private static final Pattern TAG_PATTERN = Pattern.compile("tag:(\"[^\"]+\"|\\S+)", Pattern.CASE_INSENSITIVE);

    private int pid = -1;
    private String tag;
    private String searchText;
    private int searchTextAsInt = -1;

    public SearchCriteria(CharSequence inputQuery) {

        // check for the "pid" keyword

        StringBuilder query = new StringBuilder(StringUtil.nullToEmpty(inputQuery));
        Matcher pidMatcher = PID_PATTERN.matcher(query);
        if (pidMatcher.find()) {
            try {
                pid = Integer.parseInt(pidMatcher.group(1));
                query.replace(pidMatcher.start(), pidMatcher.end(), ""); // detach
                // from
                // search
                // string
            } catch (NumberFormatException ignore) {
            }
        }

        // check for the "tag" keyword

        Matcher tagMatcher = TAG_PATTERN.matcher(query);
        if (tagMatcher.find()) {
            tag = tagMatcher.group(1);
            if (tag.startsWith("\"") && tag.endsWith("\"")) {
                tag = tag.substring(1, tag.length() - 1); // detach quotes
            }
            query.replace(tagMatcher.start(), tagMatcher.end(), ""); // detach
            // from
            // search
            // string
        }

        // everything else becomes a search term
        searchText = query.toString().trim();

        try {
            searchTextAsInt = Integer.parseInt(searchText);
        } catch (NumberFormatException ignore) {
        }

    }

    public boolean isEmpty() {
        return pid == -1 && TextUtils.isEmpty(tag) && TextUtils.isEmpty(searchText);
    }

    public boolean matches(LogLine logLine) {

        // consider the criteria to be ANDed
        if (!checkFoundPid(logLine)) {
            return false;
        }
        if (!checkFoundTag(logLine)) {
            return false;
        }
        return checkFoundText(logLine);
    }

    private boolean checkFoundText(LogLine logLine) {
        return TextUtils.isEmpty(searchText)
                || (searchTextAsInt != -1 && searchTextAsInt == logLine.getProcessId())
                || (logLine.getTag() != null && StringUtil.containsIgnoreCase(logLine.getTag(), searchText))
                || (logLine.getLogOutput() != null && StringUtil.containsIgnoreCase(logLine.getLogOutput(), searchText));
    }

    private boolean checkFoundTag(LogLine logLine) {
        return TextUtils.isEmpty(tag)
                || (logLine.getTag() != null && StringUtil.containsIgnoreCase(logLine.getTag(), tag));
    }

    private boolean checkFoundPid(LogLine logLine) {
        return pid == -1 || logLine.getProcessId() == pid;
    }

}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/loginfo/util/SearchCriteria.kt
package com.didichuxing.doraemonkit.kit.loginfo.util

import android.text.TextUtils
import com.didichuxing.doraemonkit.kit.loginfo.LogLine
import com.didichuxing.doraemonkit.kit.loginfo.util.StringUtil.containsIgnoreCase
import java.util.regex.Pattern


class SearchCriteria(inputQuery: CharSequence?) {
    companion object {
        private val PID_PATTERN = Pattern.compile("pid:(\\d+)", Pattern.CASE_INSENSITIVE)
        private val TAG_PATTERN = Pattern.compile("tag:(\"[^\"]+\"|\\S+)", Pattern.CASE_INSENSITIVE)
    }

    private var pid = -1
    private var tag: String? = null
    private var searchText: String
    private var searchTextAsInt = -1

    val isEmpty: Boolean
        get() = pid == -1 && TextUtils.isEmpty(tag) && TextUtils.isEmpty(searchText)

    init {
        val query = StringBuilder(StringUtil.nullToEmpty(inputQuery))
        val pidMatcher = PID_PATTERN.matcher(query)
        if (pidMatcher.find()) {
            try {
                pid = pidMatcher.group(1)?.toIntOrNull() ?: -1
                query.replace(pidMatcher.start(), pidMatcher.end(), "")
            } catch (e: Throwable) {
            }
        }

        val tagMatcher = TAG_PATTERN.matcher(query)
        if (tagMatcher.find()) {
            tag = tagMatcher.group(1)
            tag?.run {
                if (startsWith("\"") && endsWith("\"")) {
                    tag = this.substring(1, this.lastIndex)
                }
            }
            query.replace(tagMatcher.start(), tagMatcher.end(), "")
        }

        searchText = query.toString().trim()

        try {
            searchTextAsInt = searchText.toInt()
        } catch (ignore: NumberFormatException) {
        }
    }

    fun matches(logLine: LogLine): Boolean {
        if (!checkFoundPid(logLine))
            return false

        if (!checkFoundTag(logLine))
            return false

        return checkFoundText(logLine)
    }

    private fun checkFoundPid(logLine: LogLine): Boolean {
        return pid == -1 || logLine.processId == pid
    }


    private fun checkFoundTag(logLine: LogLine): Boolean {
        return (TextUtils.isEmpty(tag)
                || logLine.tag != null && containsIgnoreCase(logLine.tag, tag))
    }


    private fun checkFoundText(logLine: LogLine): Boolean {
        return (TextUtils.isEmpty(searchText)
                || searchTextAsInt != -1 && searchTextAsInt == logLine.processId
                || logLine.tag != null && containsIgnoreCase(logLine.tag, searchText)
                && containsIgnoreCase(logLine.logOutput, searchText))

    }
}